src/VisualStudio/CSharp/Impl/ProjectSystemShim/CSharpProjectShim.OptionsProcessor.cs

```diff
@@ -0,0 +1,288 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.VisualStudio.LanguageServices.CSharp.ProjectSystemShim.Interop;
+using Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.CSharp.ProjectSystemShim
+{
+    internal partial class CSharpProjectShim
+    {
+        private class OptionsProcessor : VisualStudioProjectOptionsProcessor
+        {
+            private readonly VisualStudioProject _visualStudioProject;
+
+            private readonly object[] _options = new object[(int)CompilerOptions.LARGEST_OPTION_ID];
+            private string _mainTypeName;
+            private OutputKind _outputKind;
+
+            public OptionsProcessor(VisualStudioProject visualStudioProject, HostWorkspaceServices workspaceServices)
+                : base(visualStudioProject, workspaceServices)
+            {
+                _visualStudioProject = visualStudioProject;
+            }
+            
+            public object this[CompilerOptions compilerOption]
+            {
+                get
+                {
+                    return _options[(int)compilerOption];
+                }
+
+                set
+                {
+                    if (object.Equals(_options[(int)compilerOption], value))
+                    {
+                        return;
+                    }
+
+                    _options[(int)compilerOption] = value;
+                    UpdateProjectForNewHostValues();
+                }
+            }
+
+            protected override CompilationOptions ComputeCompilationOptionsWithHostValues(CompilationOptions compilationOptions, IRuleSetFile ruleSetFileOpt)
+            {
+                IDictionary<string, ReportDiagnostic> ruleSetSpecificDiagnosticOptions = null;
+
+                // Get options from the ruleset file, if any, first. That way project-specific
+                // options can override them.
+                ReportDiagnostic? ruleSetGeneralDiagnosticOption = null;
+                
+                // TODO: merge this core logic back down to the base of OptionsProcessor, since this should be the same for all languages. The CompilationOptions
```

Comment:
> // TODO: [](start = 15, length = 9)

What is preventing us from doing that? (add reasoning to the comment, or link an issue)

Reply:
This was more a general observation that isn't new to this -- we already had this logic duplicated. I suspect there's a good chance this will get deleted before we ever get to the TODO.